### DISCUSSION TITLE
What problem do quadlets actually solve?

### DISCUSSION THREAD

**Discussion by dnv** [containers/podman#20305] [General]
I have several podman containers running via systemd services generated by 'podman generate systemd --new'. They work well, the unit files are easy to read and they are quite portable and easy to use, maintain and deploy using git or whatever version control you use.

Now I am reading that "generate systemd" is being deprecated in favour of quadlets. I took a look at some examples and I really don't get what problem they are supposed to be solving. Sure, yeah, instead of a long podman run command line inside of a systemd unit file you now have a dedicated Container section. And? Why?

I seem to be losing the actual automatic generative part of "podman generate systemd --new" that allows me to instantly capture a running container's configuration and then run it as a service in perpetuity and easily port the configuration elsewhere and now have to write things by hand again and for this loss I'm gaining... what exactly?

**Comment by vrothberg** [comment]
Thanks for reaching out, @dnv.

(Copied from https://github.com/containers/podman/discussions/20218#discussioncomment-7173589):

Previously, the worklflow roughly looked as follows:

 1.  Create a container on the CLI
 2.  Run podman generate systemd
 3.  Install the unit file(s)
 4.  Reload the daemon + Start

After an update to Podman, the entire process had to be repeated in case there were fixes to podman generate systemd.

With Quadlet, the workflow is reduced to installing the files and reloading the daemon and starting the service. Services may also start automatically on boot. After updates, nothing has to be changed or done by the user. The underlying units are transparently generated by Quadlet.

That means the responsibility of fixing potential issues in the generated units is not on the user.

Note that there are no plans to remove podman-generate-systemd.  Deprecation in this case means that podman-generate-systemd will not receive new features but only bug fixes.  All new features go into Quadlet.

**Reply by dnv** [reply]
> After an update to Podman, the entire process had to be repeated in case there were fixes to podman generate systemd.

This is the part I don't get. Why would this happen? The systemd unit files already exist, as I see it, the only thing that can break them and make them stop working are changes to systemd itself (or alternatively, massive changes to "podman" command-line options used to launch/destroy containers), not to the tool used to originally generate the unit files.

**Reply by vrothberg** [reply]
`podman-generate-systemd` generates the unit files.  When fixing a bug (or generally improving the code) in the generator, the units files have to be regenerated.  And that is the responsibility of the user.  The user experience left quite some space for improvement.

With Quadlet, the generation of the units files is entirely hidden from users.  It's a more declarative approach similar to Kubernetes YAML or Compose.  Upon boot or restart of the systemd daemon, Quadlet will be called by systemd, parse the Quadlet files (e.g., a `.container` file) and generate the systemd units off that.  That means that updates and fixes to Quadlet are consumed transparently; users do not have to do anything.

Quadlet files can be shared with confidence, knowing that they'll be supported and updated transparently.  Sharing the output of `podman-generate-systemd` does not have this guarantee.

**Reply by Pamalosebi** [reply]
Well... Compose files can be shared with confidence too. Many projects do.
From them you generate your containers and then you generate your service files.
Also, feels more transparent then Quadlets.

**Reply by vrothberg** [reply]
You're free to use whatever floats your boat :^)

> Well... Compose files can be shared with confidence too. Many projects do.

Absolutely.  That's the beauty of declarative approaches such as Compose, K8s YAML and Quadlet.

In case you have a workload that would require running multiple containers (and pods) and all of that in systemd, I would recommend using Quadlet's `.kube` files which allows for running Kubernetes workloads with Quadlet+Podman inside of systemd. @ygalblum wrote a nice article about that feature: https://www.redhat.com/sysadmin/multi-container-application-podman-quadlet

> Also, feels more transparent then Quadlets.

Looking at it from a software-vendor perspective, I prefer "my" users not having to worry about updating all the systemd units all the time.  Once a Quadlet file is written, Quadlet+Podman takes care of everything else.  If we fix a bug, the fixes are consumed transparently without any user having to worry.

**Reply by ruben-symbiose** [reply]
Generating systemd files was nice, but quadlets has been so much nicer for my use cases. I really enjoy declaring my config once with an option to auto update the container :+1:
